examples/step-19/step-19.cc

```diff
@@ -0,0 +1,728 @@
+    system_rhs.reinit(dof_handler.n_dofs());
+
+    constraints.clear();
+    DoFTools::make_hanging_node_constraints(dof_handler, constraints);
+
+    VectorTools::interpolate_boundary_values(dof_handler,
+                                             BoundaryIds::cathode,
+                                             Functions::ConstantFunction<dim>(
+                                               -Constants::V0),
+                                             constraints);
+    VectorTools::interpolate_boundary_values(dof_handler,
+                                             BoundaryIds::anode,
+                                             Functions::ConstantFunction<dim>(
+                                               +Constants::V0),
+                                             constraints);
+    constraints.close();
+
+    DynamicSparsityPattern dsp(dof_handler.n_dofs());
+    DoFTools::make_sparsity_pattern(dof_handler,
+                                    dsp,
+                                    constraints,
+                                    /*keep_constrained_dofs = */ false);
+    sparsity_pattern.copy_from(dsp);
+
+    system_matrix.reinit(sparsity_pattern);
+  }
+
+
+  // @sect4{The <code>CathodeRaySimulator::assemble_system</code> function}
+
+  template <int dim>
+  void CathodeRaySimulator<dim>::assemble_system()
+  {
+    system_matrix = 0;
+    system_rhs    = 0;
+
+    const QGauss<dim> quadrature_formula(fe.degree + 1);
+
+    FEValues<dim> fe_values(fe,
+                            quadrature_formula,
+                            update_values | update_gradients |
+                              update_quadrature_points | update_JxW_values);
+
+    const unsigned int dofs_per_cell = fe.dofs_per_cell;
+
+    FullMatrix<double> cell_matrix(dofs_per_cell, dofs_per_cell);
+    Vector<double>     cell_rhs(dofs_per_cell);
+
+    std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      {
+        cell_matrix = 0;
+        cell_rhs    = 0;
+
+        fe_values.reinit(cell);
+
+        for (const unsigned int q_index : fe_values.quadrature_point_indices())
+          for (const unsigned int i : fe_values.dof_indices())
+            {
+              for (const unsigned int j : fe_values.dof_indices())
+                cell_matrix(i, j) +=
+                    (fe_values.shape_grad(i, q_index) * // grad phi_i(x_q)
+                        fe_values.shape_grad(j, q_index) * // grad phi_j(x_q)
+                        fe_values.JxW(q_index));           // dx
+            }
+
+        if (particle_handler.n_particles_in_cell(cell) > 0)
+          {
+            std::vector<Point<dim>>     particle_positions;
+
+            const typename Particles::ParticleHandler<
+            dim>::particle_iterator_range particles_in_cell =
+                particle_handler.particles_in_cell(cell);
+
+            const unsigned int n_particles_in_cell =
```

Comment:
See my comment further below in the `move_particles` function. Usually I would not worry about optimizations for a simple tutorial, but since this code will likely be copied into other codes we can optimize this the same way I mentioned below. This would also allow to create `particle_positions` immediately with the correct size and save another line).